core/lib/Cardano/Script.hs

```diff
@@ -87,15 +89,59 @@ keyHashFromBytes bytes
 -- | Validate 'Script'
 --
 -- @since 3.0.0
-validateScript :: Script -> Bool
-validateScript (RequireSignatureOf _) = True
-validateScript (RequireAllOf content) =
-    not (L.null content) && L.all ((== True) . validateScript) content
-validateScript (RequireAnyOf content) =
-    not (L.null content) && L.all ((== True) . validateScript) content
-validateScript (RequireMOf m content) =
-    m > 0 && L.length content >= fromIntegral (toInteger m) &&
-    L.all ((== True) . validateScript) content
+validateScript :: Script -> Either InvalidScriptError ()
+validateScript (RequireSignatureOf (KeyHash bytes)) =
+    case keyHashFromBytes bytes of
+        Just _ -> Right ()
+        Nothing -> Left WrongKeyHash
+validateScript (RequireAllOf content) = do
+    when (L.null content) $
+        Left EmptyList
+    scanContent content
+validateScript (RequireAnyOf content) = do
+    when (L.null content) $
+        Left EmptyList
+    scanContent content
+validateScript (RequireMOf m content) = do
+    when (m == 0) $
+        Left MZero
+    when (length content < fromInteger (toInteger m) ) $
+        Left ListTooSmall
+    scanContent content
+
+scanContent :: [Script] -> Either InvalidScriptError ()
```

Comment:
Isn't that simply `traverse validateScript` ?

Reply:
probably -> https://twitter.com/impurepics/status/1312158622302924801